### DISCUSSION TITLE
Problemi con caddy 2

### DISCUSSION THREAD

**Discussion by blakdog1978** [indomus/forum#1707] [HUB: Home Assistant]
Buongiorno, stavo seguendo la guida per raggiungere home assistant dietro nat 
https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-via-reti-provider-che-usino-nat/
Sono riuscito a far tutto arrivo per l'appunto alla parte di caddy e precisamente faccio tutto giusto fino a quando bisogna controllare se si raggiunge il dominio senza porta cioè  https://miodominio.duckdns.org e lo raggiungo alla fine però quando cerco di connettermi con il mio homeassistant non riesco cioè https://miodominio.duckdns.org:9000 non si connette troppo tempo per rispondere ....
Potete aiutarmi?

**Comment by snakuzzo** [comment]
Quindi non hai fatto tutto giusto.
Ricontrolla tutto da zero

**Comment by blakdog1978** [comment]
Ho già ricontrollato due volte e non mi cambia il risultato riesco ad accedere ad homeassistant col dominio http: ma non con https: nei log non ci sono errori sulla classe degli IP ho impostato 127.0.0.1 non riesco a capire...

**Reply by snakuzzo** [reply]
Se ti va in timeout sulla 9000 hai un problema di rete. O hai sbagliato il port forwarding, o hai sbagliato la configurazione del container di caddy, oppure non puoi raggiungere il tuo router da remoto sulla 9000.

**Comment by blakdog1978** [comment]
Ho aperto la 9000 come la 443, se caddy scarica i certificati dovrebbe essere aperta la 443, di conseguenza anche la 9000

**Reply by snakuzzo** [reply]
Posta tutte le configurazioni..perché altrimenti è difficile aiutarti.
Usa il markdown e cortesemente fai reply sotto l'ultimo messaggio e non come hai fatto ora creando un nuovo thread

**Reply by blakdog1978** [reply]
Appena sono al PC provo grazie

**Reply by blakdog1978** [reply]
caddy file
```
(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
}

https://xxxxxxxxxx.duckdns.org { 
  respond "Saluti dal Crew inDomus"
  file_server
}
https://xxxxxxxxxx.duckdns.org:9000 { 
  import https_header
  reverse_proxy http://192.168.8.208:8123
}
```

lato home assistant
```
http:
  use_x_forwarded_for: true
  trusted_proxies: 
    - 127.0.0.1   
```
    
registro di caddy
```
[cont-init.d] 00-banner.sh: exited 0.
[cont-init.d] 01-log-level.sh: executing... 
Log level is set to INFO
[cont-init.d] 01-log-level.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
INFO: Prepare Caddy...
INFO: Use built-in Caddy
v2.5.1 h1:bAWwslD1jNeCzDa+jDCNwb8M3UJ2tPa8UZFFzPVmGKs=
INFO: Prepare Caddyfile...
INFO: Caddyfile found at /config/caddy/Caddyfile
INFO: Run Caddy...
{"level":"info","ts":1653563436.3943448,"msg":"using provided configuration","config_file":"/config/caddy/Caddyfile","config_adapter":""}
{"level":"warn","ts":1653563436.3977735,"msg":"Caddyfile input is not formatted; run the 'caddy fmt' command to fix inconsistencies","adapter":"caddyfile","file":"/config/caddy/Caddyfile","line":2}
{"level":"info","ts":1653563436.4007163,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
{"level":"info","ts":1653563436.4013503,"logger":"http","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443}
{"level":"info","ts":1653563436.401618,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
{"level":"info","ts":1653563436.4014447,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x4000434d20"}
{"level":"info","ts":1653563436.402005,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv1"}
{"level":"info","ts":1653563436.409065,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/ssl/caddy"}
{"level":"info","ts":1653563436.4097788,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["xxxxxxxxxxx.duckdns.org"]}
{"level":"info","ts":1653563436.4124331,"msg":"autosaved config (load with --resume flag)","file":"/data/caddy/autosave.json"}
{"level":"info","ts":1653563436.412707,"msg":"serving initial configuration"}
{"level":"info","ts":1653563436.4125466,"logger":"tls","msg":"finished cleaning storage units"}
{"level":"info","ts":1653563436.4131463,"logger":"watcher","msg":"watching config file for changes","config_file":"/config/caddy/Caddyfile"}
```

**Reply by snakuzzo** [reply]
Allora...dalla config di Caddy mi sembra ok.
Visto che hai pubblicato il tuo FQDN (che ti consiglierei di cancellare), mi sono permesso di provare una telnet...ed effettivamente sulla 443 è ok...sulla 9000 invece va in timeout.
Mi fai vedere come hai creato il container di caddy ?

Fammi vedere il contenuto del file `docker-compose.yaml` oppure il comando docker che hai usato nel caso in cui tu non abbia utilizzato `docker-compose`

e fammi vedere pure l'output del comando

```
docker ps -a
```

altra cosa...sicuro di aver configurato correttamente il port forwarding sul router ?

```
TCP 9000 -> TCP 9000 ip raspberry 
```

**Reply by blakdog1978** [reply]
scusa ma non so dove trovare il file `docker-compose.yaml`
e non riesco a copiare dal terminale il risultato del comando,
io uso hassos su respbarry

**Reply by snakuzzo** [reply]
Allora...innanzi tutto ti chiedo scusa perché non avevo capito che eri dietro NAT (non ho letto con attenzione il primo messaggio). Stai usando quindi la VM su GCP? In tal caso può essere che non hai creato correttamente la regola della 9000 con iptables sulla VM remota oppure non hai configurato/assegnato correttamente la regola firewall su GCP.

Detto ciò...se usi HassOS, stai usando l'addon di caddy2, giusto?

**Reply by blakdog1978** [reply]
Si sto usando l'addon

**Reply by blakdog1978** [reply]
Ho ricontrollato la guida ma io vedo solo regole - udp 1194 e tcp 8123  e  iptables su porta tcp 443 e tcp 8123 
l'unica cosa che ho fatto di diverso è abilitare nella VM di GCP anche il traffico HTTP oltre a HTTPS

**Reply by snakuzzo** [reply]
Sì ma perdonami...tu la guida devi pure cercare di capirla...non solo fare copia-incolla.
Le guide sono anche spunti per cercare di imparare, e non dei semplici manuali di istruzioni.
Secondo te come fai a raggiungere la porta 9000 del tuo server se non crei la regola iptables sulla VM remota ?

**Reply by blakdog1978** [reply]
dovrei quindi creare una regola del genere ? `iptables -t nat -A PREROUTING -p tcp --dport 9000 -j DNAT --to-destination 10.138.0.6:9000`

**Reply by blakdog1978** [reply]
non so se ho capito ma ho provato ad aggiungere una regola firewall sulla porta tcp 9000 come quelle in guida e anche con iptables come quella che ho sritto sopra ma ancora non riesco a raggiungere home assistant

**Reply by snakuzzo** [reply]
Ma in locale ti funziona ?
Considera che se punti iplocale:9000 deve risponderti

**Reply by blakdog1978** [reply]
no non funziona neanche in locale

**Reply by blakdog1978** [reply]
mi da questa risposta in locale : Il client ha inviato una richiesta HTTP a un server HTTPS.

**Reply by snakuzzo** [reply]
A parte che non funzionerà mai se punti in HTTP se hai esposto in HTTPS.
Ma quello che ti chiedevo era semplicemente verificare che ci sia connettività.

Da un PC qualsiasi nella tua rete..

```
telnet ip 9000
```

**Reply by blakdog1978** [reply]
fatto sembra ci sia connettività non da nessun errore

**Reply by snakuzzo** [reply]
Allora è sbagliata la regola di routing sulla VM su GCP

**Reply by snakuzzo** [reply]
Oppure è sbagliata la regola firewall associata alla VM su GCP

**Reply by snakuzzo** [reply]
dalla VM su GCP prova a fare la stessa telnet verso l'IP che il server VPN assegna al Raspberry

es.

```
telnet 10.8.0.2 9000
```

**Reply by blakdog1978** [reply]
dal terminal della VM mi da comando non trovato

**Reply by blakdog1978** [reply]
ok ho risolto questa è la risposta `Trying 10.119.222.2...
Connected to 10.119.222.2.
Escape character is '^]'.
`

**Reply by snakuzzo** [reply]
Quindi ti confermo che il problema o è il firewall lato GCP o la regola su iptables

**Reply by blakdog1978** [reply]
la regola è così
direzione - entrata
azioni in caso di corrispondenza - consenti
tag di destinazione specificati - my-host
intervalli IPv4- 0.0.0.0/0
secondo filtro di origine - nessuno
protocolli e porte
protocolli e porte specificati - tcp 9000

**Reply by snakuzzo** [reply]
la regola firewall deve essere tipo così

![1](https://user-images.githubusercontent.com/7205371/170829929-47f0b619-c523-4c93-a0c0-32bc5ebde1bd.png)

poi prendi il target tag e lo utilizzi sulla VM

![2](https://user-images.githubusercontent.com/7205371/170829952-9023f681-37b3-45c8-9202-6acfb9b57b39.png)

Poi sulla VM con iptables gestisci il routing...

```
sudo iptables -t nat -A PREROUTING -p tcp --dport 9000 -j DNAT --to-destination 10.8.0.2:9000
sudo iptables -t nat -A POSTROUTING -j MASQUERADE
```

**Reply by blakdog1978** [reply]
Prima di tutto ti ringrazio della tua disponibilità e della tua pazienza...
Ho fatto ciò che mi hai suggerito ma il risultato non cambia quando provo ad accederea https://casamia.duckdns.org:9000 mi da connessione rifiutata da ........
Ora dopo aver seguito la guida e i tuoi consigli mi trovo ad utilizzare questi tag di rete

![cattura 1](https://user-images.githubusercontent.com/36457031/170859122-ffe06b97-66ca-4171-b739-ce7691fd3dec.PNG)

ed ad avere tutte queste regole firewall 

![Cattura 2](https://user-images.githubusercontent.com/36457031/170859270-89d4fa7f-abed-406f-b7da-144924ad48eb.PNG)

Oggi aprendo la console della VM di GCP mi ha dato questo messaggio`Valuta l'aggiunta del ruolo IAM Utente del tunnel con protezione IAP per iniziare a utilizzare Cloud IAP per l'inoltro TCP e ottenere migliori prestazioni.`
Portebbe centrare ?

**Reply by snakuzzo** [reply]
Non c'entra niente.
Ma hai visto bene i tag che hai creato e quelli che hai assegnato alla VM ? A parte che hai creato tag diversi con inolto porte uguale...hai fatto solo casino. Creati un tag solo con le regole che ti servono...non è che devi fare esattamente come ti ho fatto vedere io, altrimenti dimostri che fai copia-incolla senza riflettere.
Detto ciò...secondo te come fa a funzionare se il tag utilizzato per la 9000 non l'hai assegnato alla VM ?
Ti sei accorto di questo? immagino di no

**Reply by blakdog1978** [reply]
Ed ora nei log di hassistant ho trovato questo errore:
```
Logger: homeassistant.components.http.forwarded
Source: components/http/forwarded.py:123
Integration: HTTP ([documentation](https://www.home-assistant.io/integrations/http), [issues](https://github.com/home-assistant/home-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+http%22))
First occurred: 11:50:21 (2 occurrences)
Last logged: 11:50:22

Received X-Forwarded-For header from an untrusted proxy 192.168.8.208``
```
192.168.8.208 è IP di homeassistant

Come è possibile se caddy è installato sulla stesso computer non dovrebbe essere 127.0.0.1 ?
E se si collega a https://casamia.duckdns.org non dovrebbe essere giusta la classe ip impostata in home assistant ?

Dunque nalla configurazione di home assistant dovrei mettere una classe IP ti 192.168.8.0/24 ? O solo 192.168.8.208?

**Reply by snakuzzo** [reply]
Non può essere 127.0.0.1 se usi i container...perché 127.0.0 1 è la loopback del singolo container.
Tu metti l'indirizzo che trovi nel log o comunque, se vuoi, una subnet che lo contenga

**Reply by blakdog1978** [reply]
sto usando l'addon caddy 2 è considerato un container ?

> Non c'entra niente. Ma hai visto bene i tag che hai creato e quelli che hai assegnato alla VM ? A parte che hai creato tag diversi con inolto porte uguale...hai fatto solo casino. Creati un tag solo con le regole che ti servono...non è che devi fare esattamente come ti ho fatto vedere io, altrimenti dimostri che fai copia-incolla senza riflettere. Detto ciò...secondo te come fa a se il tag utilizzato per la 9000 non l'hai assegnato alla VM ? Ti sei accorto di questo? immagino di no

Dunque devo puntare my-host come le altre due regole giusto ?

![cattura3](https://user-images.githubusercontent.com/36457031/170975240-18c92c6a-4bcb-4e7b-a2dc-da35c96bbf73.PNG)

**Reply by snakuzzo** [reply]
ogni addon è un container.
Visto che hai creato una regola diversa...devi assegnare semplicemente quel tag `homeassistant3` alla VM.
Vai in edit sulla VM e vedrai che potrai farlo

**Reply by blakdog1978** [reply]
cosi 
![cattura 4](https://user-images.githubusercontent.com/36457031/170978677-2c7580a6-e015-4a14-af85-5ebf73c10203.PNG)
?

**Reply by snakuzzo** [reply]
sì

**Reply by blakdog1978** [reply]
Ma la regola firewall come destinazione deve puntare my-host o host-https o altro ?

**Reply by snakuzzo** [reply]
deve essere uguale a quella che ti funziona...ovvero la 443 se non ricordo male.
il target tag deve corrispondere ad uno dei tag utilizzati per la tua VM

**Reply by blakdog1978** [reply]
Non riesco ancora a farlo andare... sto provando a vedere quali regole iptables siano presenti sulla VM con il comando `iptables -L`
il risultato è questo
```
root@homeassistant:/home/rvrgnn# sudo iptables -L


Chain` INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```


Sembra che non ce ne siano....
eppure il comando per salvarle l'ho dato...

**Reply by snakuzzo** [reply]
Quelle regole di PREROUTING non le vedi solo con `-L`, ma così...

```
sudo iptables -L -n -t nat
```

Es. la mia VM su GCP...

```
[snakuzzo@ha-cloud-vm ~]$  sudo iptables -L -n -t nat
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  0.0.0.0/0            0.0.0.0/0            ADDRTYPE match dst-type LOCAL
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.8.0.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.8.0.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.8.0.2:9000
...
...
```

**Reply by blakdog1978** [reply]
OK grazie questo è il mio risultato

```
root@homeassistant:/home/rvrgnn# sudo iptables -L -n -t nat
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.138.0.6:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:443 to:10.119.222.2:443
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:8123 to:10.119.222.2:8123
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
MASQUERADE  all  --  10.119.222.0/24      0.0.0.0/0            /* openvpn-nat-rule */
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0           
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0  
```

**Reply by snakuzzo** [reply]
ma quante volte hai aggiunto le regole ? Così è un bordello !!!
Tra le altre cose vedo già all'inizio che c'è un prerouting con sorgente stessa porta ma verso due IP completamente diversi

```
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.138.0.6:9000
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:9000 to:10.119.222.2:9000
```

potrebbe essere quello il problema.

Però una cosa...stiamo andando parecchio oltre. Il mio consiglio è ... riparti da una situazione pulita, perché se cominci a fare casino senza sapere bene dove metti le mani, poi perdi solo tempo e non ne esci più

**Reply by blakdog1978** [reply]
Ti ringrazio ancora per la tua pazienza sei gentilissimo, ho trovato su un forum i comandi per cancellare tutto il casino che avevo fatto, li posto così magari servirà a qualcuno...

 ```
iptables -F
  iptables -F -t mangle
  iptables -F -t nat
  iptables -X
  iptables -X -t mangle
  iptables -X -t nat
  iptables -P INPUT ACCEPT
  iptables -P OUTPUT ACCEPT
  iptables -P FORWARD ACCEPT
```
Sono i comandi per cancellare tutti ma proprio tutti gli iptables, poi ho ridato quelli giusti come da guida piu quello per la porta 9000 ed ora sono raggiungibile in https.
Se ti capita di passare da Novara avvertimi che ti pago una birra volentieri!

**Reply by snakuzzo** [reply]
Ottimo! :)
